cabal-install/src/Distribution/Client/SetupWrapper.hs

```diff
@@ -996,334 +988,408 @@ getExternalSetupMethod verbosity options pkg bt = do
-          options'' = options'{usePackageIndex = Just index}
-      case PackageIndex.lookupDependency index cabalDepName cabalDepVersion of
-        [] ->
-          dieWithException verbosity $ InstalledCabalVersion (packageName pkg) (useCabalVersion options)
-        pkgs ->
-          let ipkginfo = fromMaybe err $ safeHead . snd . bestVersion fst $ pkgs
-              err = error "Distribution.Client.installedCabalVersion: empty version list"
-           in return
-                ( packageVersion ipkginfo
-                , Just . IPI.installedComponentId $ ipkginfo
-                , options''
-                )
-
-    bestVersion :: (a -> Version) -> [a] -> a
-    bestVersion f = firstMaximumBy (comparing (preference . f))
+      copyFileVerbose verbosity customSetupHooks (i (setupHooks options))
+      rewriteFileLBS verbosity (i (setupHs options)) (buildTypeScript Hooks cabalLibVersion)
+      rewriteFileLBS verbosity (i (hooksHs options)) hooksExeScript
+    updateSetupScript cabalLibVersion bt' =
+      rewriteFileLBS verbosity (i (setupHs options)) (buildTypeScript bt' cabalLibVersion)
+
+-- | The source code for a non-Custom 'Setup' executable.
+buildTypeScript :: BuildType -> Version -> BS.ByteString
+buildTypeScript bt cabalLibVersion = "{-# LANGUAGE NoImplicitPrelude #-}\n" <> case bt of
+  Simple -> "import Distribution.Simple; main = defaultMain\n"
+  Configure
+    | cabalLibVersion >= mkVersion [3, 13, 0]
+    -> "import Distribution.Simple; main = defaultMainWithSetupHooks autoconfSetupHooks\n"
+    | cabalLibVersion >= mkVersion [1, 3, 10]
+    -> "import Distribution.Simple; main = defaultMainWithHooks autoconfUserHooks\n"
+    | otherwise
+    -> "import Distribution.Simple; main = defaultMainWithHooks defaultUserHooks\n"
+  Make -> "import Distribution.Make; main = defaultMain\n"
+  Hooks
+    | cabalLibVersion >= mkVersion [3, 13, 0]
+    -> "import Distribution.Simple; import SetupHooks; main = defaultMainWithSetupHooks setupHooks\n"
+    | otherwise
+    -> error "buildTypeScript Hooks with Cabal < 3.13"
+  Custom -> error "buildTypeScript Custom"
+
+-- | The source code for an external hooks executable, using the 'hooks-exe' library.
+hooksExeScript :: BS.ByteString
+hooksExeScript = "{-# LANGUAGE NoImplicitPrelude #-}\nimport Distribution.Client.SetupHooks.HooksExe (hooksMain); import SetupHooks; main = hooksMain setupHooks\n"
+
+installedCabalVersion
+  :: Verbosity
+  -> PackageDescription
+  -> BuildType
+  -> SetupScriptOptions
+  -> Compiler
+  -> ProgramDb
+  -> IO
+      ( Version
+      , Maybe InstalledPackageId
+      , SetupScriptOptions
+      )
+installedCabalVersion _verbosity pkg bt options' _ _
+  | packageName pkg == mkPackageName "Cabal"
+      && bt == Custom =
+      return (packageVersion pkg, Nothing, options')
+installedCabalVersion verbosity pkg _bt options' compiler progdb = do
+  index <- maybeGetInstalledPackages verbosity options' compiler progdb
```

Comment:
`installedCabalVersion` is only called once (L949) and right after `configureCompiler` which pre-emptively calls `maybeGetInstalledPackages` knowing it's going to needed here.
If I got this right, this function signature can be simplified down to

```
installedCabalVersion
  :: Verbosity
  -> InstalledPackageIndex
  -> VersionRange
  -> PackageId
  -> BuildType
  -> IO (Version, Maybe InstalledPackageId)
```

which has the advantage of not dealing with `SetupScriptOptions` mutation's. See my attempt https://github.com/haskell/cabal/pull/9735/files#diff-6ff0f0024945e11d4f23279d3d025c146ef7b28d15e29fefe73f1dab10a80325R793-R799